include/evmc/mocked_host.hpp

```diff
@@ -0,0 +1,320 @@
+// EVMC: Ethereum Client-VM Connector API.
+// Copyright 2019 The EVMC Authors.
+// Licensed under the Apache License, Version 2.0.
+#pragma once
+
+#include <evmc/evmc.hpp>
+#include <algorithm>
+#include <iterator>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace evmc
+{
+/// The string of bytes.
+using bytes = std::basic_string<uint8_t>;
+
+/// Mocked account.
+struct MockedAccount
+{
+    /// Extended value (by dirty flag) for account storage.
+    struct storage_value
+    {
+        /// The storage value.
+        bytes32 value;
+
+        /// True means this value has been modified already by the current transaction.
+        bool dirty{false};
+
+        /// Default constructor.
+        storage_value() noexcept = default;
+
+        /// Constructor.
+        storage_value(const bytes32& _value, bool _dirty = false) noexcept  // NOLINT
+          : value{_value}, dirty{_dirty}
+        {}
+    };
+
+    /// The account code.
+    bytes code;
+
+    /// The code hash. Can be a value not related to the actual code.
+    bytes32 codehash;
+
+    /// The account balance.
+    uint256be balance;
+
+    /// The account storage map.
+    std::unordered_map<bytes32, storage_value> storage;
+
+    /// Helper method for setting balance by numeric type.
+    void set_balance(uint64_t x) noexcept
+    {
+        balance = uint256be{};
+        for (std::size_t i = 0; i < sizeof(x); ++i)
+            balance.bytes[sizeof(balance) - 1 - i] = static_cast<uint8_t>(x >> (8 * i));
+    }
+};
+
+/// Mocked EVMC Host implementation.
+class MockedHost : public Host
+{
+public:
+    /// LOG record.
+    struct log_record
+    {
+        /// The address of the account which created the log.
+        address creator;
+
+        /// The data attached to the log.
+        bytes data;
+
+        /// The log topics.
+        std::vector<bytes32> topics;
+
+        /// Equal operator.
+        bool operator==(const log_record& other) const noexcept
+        {
+            return creator == other.creator && data == other.data &&
+                   topics.size() == other.topics.size() &&
+                   std::equal(topics.begin(), topics.end(), other.topics.begin());
```

Comment:
wouldn't `topics == other.topics` work here?

Reply:
I think it should. I missed the fact `std::vector` has `==`.

Reply:
Will you change this?

Reply:
This is changed.

Reply:
I might be blind, but it still uses `std::equal` with iterators?

Reply:
Yes, but these comments are attached to the outdated code fragment.